### PR TITLE
Add sidebar to plugin manager, increase search bar size

### PR DESCRIPTION
**Pull request by janfaracik**
This is a small-ish PR to add a sidebar to the plugin manager, as well as to increase the size and prominence of the search bar.

**What's changed**
* New sidebar
* Update count now displays in the sidebar
* Search bar is now larger and sticks to the top of the page
* Misc text changes

**Before**
image

**After**
image

https://user-images.githubusercontent.com/43062514/177374904-27c0541a-0aac-4c1d-b8bb-0b7822d88890.mov

### Proposed changelog entries

* Add sidebar to plugin manager, increase search bar size

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6783"></a>

### CODE REVIEW

**Comment by daniel-beck** [comment]
Previously: https://github.com/jenkinsci/jenkins/pull/4670

**Comment by janfaracik** [comment]
I'd like to merge Updates and Available in the future if there's an appetite for that, I demoed a rough implementation of it a while back https://youtu.be/fnigHDZSIqQ

File: core/src/main/resources/hudson/PluginManager/available.jelly
Comment (uhafner) [inline core/src/main/resources/hudson/PluginManager/available.jelly:34]: This is a little bit unrelated as the current behavior already shows this problems:

When I click on the side bar link I would assume that something happens immediately. However, it takes some seconds on my machine until the new view is rendered (due to the slow server query). I think it would improve the user experience if a small "loading..." text is shown in the table until the data has been received from the server. 

Additionally (some are existing problems): 
- it is not clear how many plugins have been loaded (and why only those) 
- how many plugins are available in total
- sorting is not applied after searching 
- several primary buttons

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (uhafner) [inline core/src/main/resources/hudson/PluginManager/sidepanel.jelly:99]: Shouldn't there be a Back button as well?

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (uhafner) [inline core/src/main/resources/hudson/PluginManager/sidepanel.jelly:53]: Shouldn't the badge be right after the text without a gap?

**Review by timja — COMMENTED** [review]
Looks great!

Tested locally all good to me.

Just needs ATH run through to be safe

**Review by uhafner — COMMENTED** [review]
I like the idea of replacing the side panel with a local navigation while looking at a details page! This also provides a way that the side panel has fixed height and is non-vertical-scrolling on the left, while the main panel can fill the whole screen. And if the main panel is larger, then only the main panel scrolls, not the side panel! (Though I am not sure that everybody will like that concept in general in Jenkins)

File: core/src/main/resources/hudson/PluginManager/available.jelly
Comment (timja) [reply core/src/main/resources/hudson/PluginManager/available.jelly:34]: Do you get delay on this page? I only get a delay on the installed plugins page, which is slower than it should be.

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (timja) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:99]: Ideally the `Manage Jenkins` breadcrumb should be resurrected so we don't need that.

i.e https://github.com/jenkinsci/jenkins/pull/6126 (not sure why it was closed, if there was something missing in it, cc @daniel-beck

Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:99]: Yes, it is part of the breadcrumb then it is not needed indeed.

Comment (daniel-beck) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:99]: > not sure why it was closed, if there was something missing in it

https://github.com/jenkinsci/jenkins/pull/6485#issuecomment-1175251633

File: core/src/main/resources/hudson/PluginManager/available.jelly
Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/available.jelly:34]: On this page about 4 seconds and on the installed plugins 8 seconds. Typical UI guidelines recommend a progress indicator if an operation takes more than 1 second.

Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/available.jelly:34]: (But I am running a lot on my machine right now, so this is not a good indicator)

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/available.jelly:34]: Right I certainly don't notice anything on the available page, installed is very noticeable though.

No harm to add a loading spinner though, but could be done separately.

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (janfaracik) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:53]: I've placed it on the far right so that if there are multiple sidebar items with badges they're all aligned together, happy to change it though if you feel strongly 👍

File: core/src/main/resources/hudson/PluginManager/available.jelly
Comment (janfaracik) [reply core/src/main/resources/hudson/PluginManager/available.jelly:32]: > This is a little bit unrelated as the current behavior already shows these problems:
> 
> When I click on the side bar link I would assume that something happens immediately. However, it takes some seconds on my machine until the new view is rendered (due to the slow server query). I think it would improve the user experience if a small "loading..." text is shown in the table until the data has been received from the server.
> 
> Additionally (some are existing problems):
> 
> * it is not clear how many plugins have been loaded (and why only those)
> * how many plugins are available in total
> * sorting is not applied after searching
> * several primary buttons

I've added these to my plugin manager to-do list - thanks, it's super appreciated!

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/available.jelly:32]: not sure if it's on your to-do or not, (possibly this https://github.com/jenkinsci/jenkins/pull/6783#discussion_r914917897)

but the select all / none at the bottom is very annoying.

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (timja) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:53]: I like it where it is

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (janfaracik) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:1]: Definitely 👍 If it's alright with you though I'd like to wait a bit to have a more concrete design across Jenkins before building a new component/tag for these new sorts of sidebar behaviours as they may change/be tweaked still depending on future PRs.

**Comment by daniel-beck** [comment]
This seems to work well. I especially like that on applicable views, the search field gets focused. That was always annoying. Nobody can navigate Available or even Installed without that.

Some questions:

1. Is it intentional that the update center is not getting the same/similar sidebar as plugins? It's a bit jarring to go from PM to UC and see the sidebar change to completely.
2. This is the second PR introducing a title in the sidepanel (after #6485) without even making a note of that. Is that documented and discussed somewhere? How are we going to handle the problem of pages without a sidepanel, won't it be weird when some pages have a title above the main panel, while others have it in the sidepanel?

File: war/src/main/less/pages/plugin-manager.less
Comment (uhafner) [reply war/src/main/less/pages/plugin-manager.less:125]: 👍

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:53]: No, it's not important. It depends a little bit how far away these badges are. When we have large label texts then this might not look good anymore.

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (uhafner) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:1]: I am not sure if this is a good idea. As @daniel-beck noted as well:

> This is the second PR introducing a title in the sidepanel (after https://github.com/jenkinsci/jenkins/pull/6485) without even making a note of that. Is that documented and discussed somewhere? How are we going to handle the problem of pages without a sidepanel, won't it be weird when some pages have a title above the main panel, while others have it in the sidepanel?

I think we should start very soon with creating and documenting a control for that (and maybe opening a discussion about the new concept can and should be used in plugins). Or when do you expect that others will pick up those changes for their plugins? Then we should at least notify everybody that this concept is not yet mature and should not be applied in plugins.

**Review by timja — APPROVED** [review]
Implementation, design and testing wise this looks great to me.

Update center could definitely do with a face lift, at least removing the unneeded links, but I don't think it needs to be done in this PR. I'm also not sure about if just adding the plugin manager side bar would be useful there (but it could be done)

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Item 2 in https://github.com/jenkinsci/jenkins/pull/6783#issuecomment-1176740601 needs to be answered.

**Comment by janfaracik** [comment]
> 1. Is it intentional that the update center is not getting the same/similar sidebar as plugins? It's a bit jarring to go from PM to UC and see the sidebar change to completely.

I don't have a particular opinion on this, happy to add it separately if need be. One of the things I wanted to do eventually is decrease how much that page is shown (https://youtu.be/fnigHDZSIqQ?t=840) and instead to have download information be inline in the table.

> 2. This is the second PR introducing a title in the sidepanel (after [[JENKINS-68282] New design for configure project page #6485](https://github.com/jenkinsci/jenkins/pull/6485)) without even making a note of that. Is that documented and discussed somewhere? How are we going to handle the problem of pages without a sidepanel, won't it be weird when some pages have a title above the main panel, while others have it in the sidepanel?

End goal I think there would be to have the header above the side panel if the page has one (and an optional header above main content if it needs it) just right now a lot of pages haven't been updated/looked at it so forcing a title on them might be odd/waste vertical space. 

One thing I'm keen on is having actions grouped in the app-bar, e.g. like the log recorders page/new project manage page, and then using the sidebar solely for navigation between child pages. Still a long way to go though.

**Review by daniel-beck — DISMISSED** [review]
There are several mostly unrelated changes combined in this PR. Moving the "tabs" for the various views of the plugin manager to the sidepanel seems fine. The increased prominence of the search bar is great, as it is the only easy way to interact with these pages.

Using a plain HTML tag as a separator in the sidepanel is potential trouble as a new UI element:

* If this gets widely adopted by plugins for their sidepanels, we might end up needing to retain compatibility for a specific tag when a different approach might be better.
* How does this interact with extensible lists of tasks (~ `Actionable`) and how would they add separators for sidepanel items?
* Should separators show in context menus of `ModelObjectWithContextMenu`? If not, why not?

Considering the experience with section titles in _Configure System_ that end up visually grouping some items without a separate title nonsensically (#2744), consider instead looking into defining task groups, with the boundary between two groups indicated by a separator.


With regards to the title in the sidepanel, given

> End goal I think there would be to have the header above the side panel if the page has one

I'd like to understand the benefits of this change first. AFAICT, this probably affects hundreds views in at least dozens of plugins, and this PR doesn't even convert all views in Jenkins to investigate feasibility.

Both of these changes need developer documentation on how _and when_ to use these (the latter in particular is currently lacking from Design Library).

Let's please not introduce new UI paradigms that need views in hundreds of plugins to be adapted for consistency without accompanying documentation as an unmentioned change in a mostly unrelated PR.

The change to move the title has JEP-level impact, so should go through a wider contributor review first (even if not doing a full JEP). This doesn't only impact core, but tons of plugins. The separator isn't quite as bad, but how to apply it more widely and how it interacts with existing Jenkins UI needs to be determined before this is merged.

**Comment by janfaracik** [comment]
> There are several mostly unrelated changes combined in this PR. Moving the "tabs" for the various views of the plugin manager to the sidepanel seems fine. The increased prominence of the search bar is great, as it is the only easy way to interact with these pages.
> 
> Using a plain HTML tag as a separator in the sidepanel is potential trouble as a new UI element:
> 
> * If this gets widely adopted by plugins for their sidepanels, we might end up needing to retain compatibility for a specific tag when a different approach might be better.
> * How does this interact with extensible lists of tasks (~ `Actionable`) and how would they add separators for sidepanel items?
> * Should separators show in context menus of `ModelObjectWithContextMenu`? If not, why not?
> 
> Considering the experience with section titles in _Configure System_ that end up visually grouping some items without a separate title nonsensically (#2744), consider instead looking into defining task groups, with the boundary between two groups indicated by a separator.
> 
> With regards to the title in the sidepanel, given
> 
> > End goal I think there would be to have the header above the side panel if the page has one
> 
> I'd like to understand the benefits of this change first. AFAICT, this probably affects hundreds views in at least dozens of plugins, and this PR doesn't even convert all views in Jenkins to investigate feasibility.
> 
> Both of these changes need developer documentation on how _and when_ to use these (the latter in particular is currently lacking from Design Library).
> 
> Let's please not introduce new UI paradigms that need views in hundreds of plugins to be adapted for consistency without accompanying documentation as an unmentioned change in a mostly unrelated PR.
> 
> The change to move the title has JEP-level impact, so should go through a wider contributor review first (even if not doing a full JEP). This doesn't only impact core, but tons of plugins. The separator isn't quite as bad, but how to apply it more widely and how it interacts with existing Jenkins UI needs to be determined before this is merged.

Writing up some guidelines for this at the moment for the Design Library - will post it here when its in draft. Cheers

File: core/src/main/resources/hudson/PluginManager/advanced.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginManager/advanced.jelly:31]: Is `Detail - Category` a reasonable presentation universal across languages Jenkins is translated to?

Comment (janfaracik) [reply core/src/main/resources/hudson/PluginManager/advanced.jelly:31]: That's a good question and something I haven't got a concrete answer to. I've given it a quick Google but haven't found any articles on the topic, so I've looked at how various sites behave with right-to-left languages:

Google Search - Doesn't change the hierarchy of the title (e.g. Jenkins - Google)
Wikipedia - Does change the hierarchy of the title (e.g. Wikipedia - Jenkins)
Google Translate - Does change the hierarchy of the title

**Comment by timja** [comment]
@daniel-beck have you got any thought on @janfaracik's responses both here and in Gitter?

https://gitter.im/jenkinsci/ux-sig?at=62d85ab4b408e830dc24aca0

Is your review still blocking? What can we do to move it forward here.

Cheers

**Comment by daniel-beck** [comment]
FTR I responded in Gitter.

File: core/src/main/resources/hudson/model/Messages.properties
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Messages.properties:285]: Unnecessary

Comment (timja) [reply core/src/main/resources/hudson/model/Messages.properties:285]: outdated diff?

**Comment by janfaracik** [comment]
Thanks for all the help on this @timja :) Super appreciated.

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginManager/sidepanel.jelly:27]: Why was the Groovy file changed to Jelly?

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginManager/sidepanel.jelly:36]: Why is this still a custom sidepanel implementation?

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:36]: I could do that, (and I have implemented it locally)

but we would lose support for the update count badge.

image

image

any thoughts?

Comment (daniel-beck) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:36]: As usual, if this is going to be a thing we do, it should be properly supported, e.g. by being added to `l:task`.

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:36]: @janfaracik what do you think shall we add to task now or temporarily remove and enhance later?

Comment (timja) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly]: I've removed it for now

**Comment by daniel-beck** [comment]
Interactive testing of this PR is a jarring experience. I don't know whether it is just because it's different from before, or whether moving the links into the sidepanel is a problem more generally.

Going to the plugin manager I am arriving on this page:

Screenshot 2022-08-19 at 11 37 05

I am now confused because there's nothing I can do here.

(Not negative feedback per se, just an observation.)

----

Additionally, the very wide input field, while looking nice, makes the browser-provided suggestions also very wide, which looks a bit weird. Probably nothing that can (or should?) be done here.

Before: before

After: after

**Comment by janfaracik** [comment]
> I am now confused because there's nothing I can do here.
> 
> (Not negative feedback per se, just an observation.)

I've never liked that the 'Updates' page was the default to be fair, 9/10 when I go to Plugins it's because I want to install a new plugin. We'll hopefully merge the 'Updates' and 'Installed' pages together one day as prototyped here https://youtu.be/fnigHDZSIqQ?t=840 which should solve that.

> Additionally, the very wide input field, while looking nice, makes the browser-provided suggestions also very wide, which looks a bit weird. Probably nothing that can (or should?) be done here.

That's odd, I don't get suggestions (at least in Safari - I'm assuming it's your past input?), not too sure how beneficial they are for that input so might be worth disabling later down the line.

**Comment by timja** [comment]
> That's odd, I don't get suggestions (at least in Safari - I'm assuming it's your past input?), not too sure how beneficial they are for that input so might be worth disabling later down the line.

I get some suggestions in Chrome which look nice:

image

Unlikely to be useful to actual users, more for developers of Jenkins and not that useful tbh.

----

I didn't manage to get any suggestions in firefox even by doing some searches and installing some plugins

**Comment by daniel-beck** [comment]
> I didn't manage to get any suggestions in firefox even by doing some searches and installing some plugins

This PR moves the search field outside of the form, so maybe your inputs are no longer remembered as part of an earlier form submission?

**Comment by timja** [comment]
Do you want something done here?

We could try disabling autocomplete if required.

——

Can you clarify what  is holding back the approval please?

**Comment by daniel-beck** [comment]
> Can you clarify what is holding back the approval please?

Available time and competing priorities. I like to review and test changes in depth. What I wrote above was initial feedback when running this PR, haven't gotten around to a more extensive test.

**Review by NotMyFault — COMMENTED** [review]
- `manage/pluginManager/available`'s bottom sticker is detached from the background, until you scroll to the bottom of the page and back up:

https://user-images.githubusercontent.com/13383509/186260345-164fddbd-bb63-464d-a735-ecfb482cda70.mp4

~~- I'd be in favor of dropping the search bar suggestions for now, if that's a specific addition of this PR, because this PR targets the sidebar and search bar _size_. We can reiterate that on a later PR, but I feel minor changes like these end up in delaying PRs longer than needed.~~

Edit: Disregard the latter, the same behavior is present on recent weeklys.

**Comment by timja** [comment]
> manage/pluginManager/available's bottom sticker is detached from the background, until you scroll to the bottom of the page and back up:

I haven't managed to reproduce this (on chrome)

Can you reliably reproduce it? and are there any specific steps

**Comment by NotMyFault** [comment]
> Can you reliably reproduce it? and are there any specific steps

Dashboard -> Manage Jenkins -> Manage Plugins -> Available plugins on a clean installation let me reproduce it with Chrome 104.

Oddly enough, I can't reproduce it with Safari or Firefox :/

**Review by NotMyFault — APPROVED** [review]
Looks fancy 🎉 

I'm fine with disregarding my previous concern, given input suggestions aren't introduced here and do already exist on the past weeklys.

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (basil) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:1]: As far as I can tell #7051 is the codification of this design pattern. Am I correct in understanding that after #7051 is merged, this PR and the Configure Project redesign will be refactored to use the new codified design pattern, at which point this feedback will be completely addressed?

Comment (janfaracik) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:1]: The configure project is updated as part of #7051. The plugin manager on the other hand has been changed (I should probably update the opening post) to get sign off from Daniel so the title of the page is now above the search bar.  This could change in the future in which case I'll definitely adopt the way #7051 does it 👍

**Comment by basil** [comment]
Wow, this looks great! I am very pleased with the result and I think this is a huge step forward. I only have one piece of optional non-blocking feedback for your consideration.

When I click on "Download progress" on the sidebar, I get to a page with "Download progress" as its heading. When I click on "Advanced settings" on the sidebar, I get to a page with "Advanced settings" as its heading. However, when I click on "Updates", "Available Plugins", or "Installed Plugins" on the sidebar, I get to a page with "Plugins" as its heading. To me, this is unexpected because I would expect the heading of the page to match the navigation link. To me, this feels like it is possibly a relic of the older implementation (where these three pieces of functionality were on a single page with three tabs) that was unintentionally carried through into the redesigned version. But whether or not it was intentional, it seems inconsistent and therefore a potential source of confusion to users. My vote would be to make the headings consistent with the navigation links in all cases. If I am wrong about something here, please feel free to correct me.

File: core/src/main/resources/hudson/PluginManager/sidepanel.jelly
Comment (basil) [reply core/src/main/resources/hudson/PluginManager/sidepanel.jelly:1]: I didn't really follow the explanation regarding the relationship between this PR and #7051, but I trust that we are making progress. My primary concern is that the design pattern codified in #7051 is consistently followed by all implementations, including the configure project page and the plugin manager page (this PR) as well as any future pages that are converted to this pattern.

The order of operations seems like it is up for debate, but my main concern is just that we reach the desired end state and not that we take any particular order of operations to get there. I would expect that it would be desirable to fully flesh out the API in #7051 before adding new implementations as in this PR, because otherwise those implementations would then need to be adjusted later to conform to the final API. But I could be wrong about this. Either way, I think we should get the API finalized and all implementations conforming before we convert more pages to this paradigm.

**Comment by timja** [comment]
>  But whether or not it was intentional, it seems inconsistent and therefore a potential source of confusion to users. My vote would be to make the headings consistent with the navigation links in all cases

I think that will be mostly addressed when https://github.com/jenkinsci/jenkins/pull/6783#discussion_r963099061 is sorted.

The title will move into the sidebar and the plugin manager pages themselves won't have a main title.

Basically the 3 tabs are being considered as part of 1 page almost.
advanced being a bit special and it could be somewhat removed with some additional refactoring out of scope of this PR.

----

If no objections I plan to merge this tomorrow towards the next weekly.

File: war/src/main/less/modules/side-panel-tasks.less
Comment (daniel-beck) [inline war/src/main/less/modules/side-panel-tasks.less:82]: Unused in this PR, only used in https://github.com/jenkinsci/jenkins/pull/7084.

File: war/src/main/less/modules/side-panel-tasks.less
Comment (daniel-beck) [inline war/src/main/less/modules/side-panel-tasks.less:90]: Improperly resolved conflict, or what is this for?

Comment (janfaracik) [reply war/src/main/less/modules/side-panel-tasks.less:90]: You're correct 👍 I've removed it as part of #7084

**Pull request merged**
